src/main/java/org/sqlite/jdbc3/JDBC3Statement.java

```diff
@@ -418,30 +424,49 @@ protected SQLException unsupported() {
         return new SQLFeatureNotSupportedException("not implemented by SQLite JDBC driver");
     }
 
+    private int executeInsert(String sql, String keys) throws SQLException {
+        int pos = sql.lastIndexOf(';');
+        int index = pos != -1 ? pos : sql.length();
+        String prefix = sql.substring(0, index);
+        String suffix = sql.substring(index);
+        sql = prefix + " RETURNING " + keys + suffix;
```

Comment:
I believe this is quite dangerous to perform such changes on the fly on the query. There is no query manipulation in this driver, and if we were to add it, it would need a proper sql parser.

My understanding from your issue was that if the user submits a query with RETURNING, the keys would be made available, else they would be ignored.

Reply:
> I believe this is quite dangerous to perform such changes on the fly on the query. There is no query manipulation in this driver, and if we were to add it, it would need a proper sql parser.

Simply add (before a possible ;) the word RETURNING and a * for all the columns or columns chosen by their name. I don't think a parser is needed to make such simple changes.
This may seem a bit ugly but I believe it is the right way to get as close as possible to the JDBC API.

> My understanding from your issue was that if the user submits a query with RETURNING, the keys would be made available, else they would be ignored.

No more precisely if the user submits a query with:
- execute(String sql, int autoGeneratedKeys) or executeUpdate(String sql, int autoGeneratedKeys) with autoGeneratedKeys == RETURN_GENERATED_KEYS
- execute(String sql, String[] colnames)
- executeUpdate(String sql, String[] colnames)
- executeLargeUpdate(String sql, String[] colnames)
Then all the keys or the one requested will be available in the getGeneratedKeys() method.

Reply:
> This may seem a bit ugly

you got the gist of it. I am not keen on proceeding on merging this PR in that state.

Reply:
OK what do you suggest?

Reply:
as i said, if a statement contains `RETURNING`, the driver could return the generated keys.

Reply:
Maybe we can parse it like this:
```
int pos = sql.indexOf(';');
int index = pos != -1 ? pos : sql.length();
StringBuilder buffer = new StringBuilder(sql.substring(0, index));
buffer.append(" RETURNING ");
buffer.append(keys);
buffer.append(sql.substring(index));
sql = buffer.toString();
```
In any case, any parser won't do much different than that?

Reply:
> as i said, if a statement contains RETURNING, the driver could return the generated keys.

**I understood that well, but why ask me to do PR if it's to answer me that?**

Your solution does not follow the JDBC API for two reasons:
- an executeUpdate cannot return a resultset.
- SQL INSERT commands must be carried out using the executeUpdate() method.

As a result, software following JDBC specifications will not work with sqlite-jdbc. It's really a shame...

Reply:
> As a result, software following JDBC specifications will not work with sqlite-jdbc. It's really a shame...

compromises have to be made, SQLite is very different from other DB engines.